Replace readdir_r

kripton opened the pull request:
This PR replaces the deprecated readdir_r with the non-deprecated readdir.
A unit test is added to make sure we don't break anything. Tested on the old readdir_r and the new code.

Open for discussions :)

kripton commented:
Oh, forgot: Fixes #1055

kripton commented:
Forgot another thing: I also wanted to revert https://github.com/OpenLightingProject/ola/pull/1168. That might lead to more warning being shown, so not tested yet. Will do so later. So please don't merge yet.

kripton commented:
Okay, Travis tests are failing due to out-of-tree builds. I'll check how other tests are handling test data

kripton commented:
Okay, I'd say this is ready for review & comments. New unit test is passing on Travis, existing tests are flaky and currently sometimes fail non macOS. Doesn't seem to be related to these changes.

peternewman commented on common/file/Util.cpp:
```diff
@@ -129,15 +129,16 @@ bool FindMatchingFiles(const string &directory,
 
-  if (readdir_r(dp, &dir_ent, &dir_ent_p)) {
-    OLA_WARN << "readdir_r(" << directory << "): " << strerror(errno);
+  errno = 0;
```
Is it worth explicitly checking that errno is non-zero after the opendir call for some weird edge-case install, before we explicitly set it to okay?

kripton replied:
True. `errno` is thread-local but checking it won't hurt. However, since `opendir` sets `errno` in case of an error (and we are checking the return value for `NULL`), it should not happen

peternewman replied:
You never quite know with some of the weird distros we support.

kripton replied:
I've investigated a bit more into this. Main points (from https://linux.die.net/man/3/errno):
```
Its value is significant only when the return value of the call indicated an error (i.e., -1 from 
most system calls; -1 or NULL from most library functions); a function that succeeds is
allowed to change errno.

Valid error numbers are all nonzero; errno is never set to zero by any system call or library function.
```

As such, checking it for `0` doesn't make sense since that would only be true if it was `0` from the beginning or was set to `0` intentionally after a last system function call.

However, setting it to `0` is required because the previous, successful function call might have left it with a value.

And checking it for a value != `0` after calling `readdir` is required since the return value of `NULL` can have two meanings:
* Either it was all okay and there are no more entries to be returned OR
* there was an error

In the first case, if errno had not been set to `0` prior to the call, I would not know if it was okay, since (according to the docs), no system function call is allowed to set errno to zero.

peternewman replied:
Given it says:

> a function that succeeds _is_ allowed to change errno.

I'm unclear what a successful opendir call and an errorno value would mean, although it seems to say it shouldn't.

peternewman replied:
Resolving as discussed in main thread now.

kripton commented on common/file/UtilTest.cpp:
```diff
@@ -90,3 +94,25 @@ void UtilTest::testFilenameFromPath() {
+
+  std::vector<std::string> files;
+
+  okay = FindMatchingFiles(std::string(TEST_SRC_DIR) + std::string("/man"),
```
Honestly, I'm not sure what you mean to say.
My plan is:
* Write a unit test for the _current_ implementation of `FindMatchingFiles` (that PASSES)
* Replace the implementation of `FindMatchingFiles`
* Verify that the unit test still PASSES to make sure the new implementation behaves identical to the old one

peternewman replied:
I think I meant `TEST_SRC_DIR` :confused: 

The plan sounds great though.

kripton replied:
The test passes on my system, so I seem it is expanded properly.
And TEST_SRC_DIR is already being used here:
* https://github.com/OpenLightingProject/ola/blob/4e5395a311bf4335440fbd40dcea19a25b99de88/common/web/SchemaParserTest.cpp#L136
* https://github.com/OpenLightingProject/ola/blob/4e5395a311bf4335440fbd40dcea19a25b99de88/common/rdm/PidStoreTest.cpp#L82

kripton commented on common/file/UtilTest.cpp:
```diff
@@ -90,3 +94,25 @@ void UtilTest::testFilenameFromPath() {
+
+  OLA_ASSERT_TRUE_MSG(okay, "FindMatchingFiles returned false");
+
+  // At the time this test was written, there were 3 files in folder "man"
+  // starting with "rdm_". If this changed, please adapt the number below
+  // Or find something better to match against
```
True, good idea. Will be done.

kripton replied:
done :)

kripton commented on configure.ac:
```diff
@@ -228,29 +228,6 @@ AC_CHECK_FUNCS([bzero gettimeofday memmove memset mkdir strdup strrchr \
                 socket strerror getifaddrs getloadavg getpwnam_r getpwuid_r \
                 getgrnam_r getgrgid_r secure_getenv])
 
-AC_MSG_CHECKING(for readdir_r deprecation)
-old_cxxflags=$CXXFLAGS
-CXXFLAGS="${CXXFLAGS} -Werror=deprecated-declarations"
-AC_CACHE_VAL(ac_cv_readdir_r_deprecation,
-  AC_COMPILE_IFELSE(
-     [AC_LANG_PROGRAM([[#include <dirent.h>
-#include <unistd.h>]],[[
-        const char* arg = "/";
-        DIR * dir;
-        struct dirent *buf, *de;
-        dir = opendir(arg);
-        readdir_r(dir,buf,&de);
-]])],
-     [ac_cv_readdir_r_deprecation=no],
-     [ac_cv_readdir_r_deprecation=yes])
-)
-CXXFLAGS=$old_cxxflags
-AC_MSG_RESULT($ac_cv_readdir_r_deprecation)
-
-AS_IF([test "x$ac_cv_readdir_r_deprecation" = xyes],
-      [CXXFLAGS="$CXXFLAGS -Wno-error=deprecated-declarations"],
-      [])
-
```
I assume not :) However, I am aware that removing the `-Wno-error=deprecated-declarations`-flag can lead to more warnings compared to before. But better be warned than masking them and not be aware

peternewman replied:
Indeed, that's why we throw all the errors and warnings by default so we know what's going to break for us next.

kripton commented:
@peternewman : All comments addressed (code changed or reasoning added ;)). Ready for a new round of review

kripton commented:
Oh, right, linting pointed out some stuff I saw just now. Will also take car of that

kripton commented:
All lint fixes done and rebased onto latest 0.10 branch

peternewman commented on common/file/UtilTest.cpp:
```diff
@@ -20,26 +20,36 @@
 #include "ola/file/Util.h"
 #include "ola/testing/TestUtils.h"
 
+#ifdef _WIN32
+const char PATH_SEPARATOR = '\\';
+#else
+const char PATH_SEPARATOR = '/';
+#endif  // _WIN32
+
```
Can we not use the definition that's already in https://github.com/OpenLightingProject/ola/blob/0.10/common/file/Util.cpp#L47-L51

Then perhaps add a test to ensure it's been set correctly?

peternewman commented on common/file/UtilTest.cpp:
```diff
@@ -90,3 +100,54 @@ void UtilTest::testFilenameFromPath() {
   OLA_ASSERT_EQ(string("baz"), FilenameFromPathOrPath("/foo/bar/baz"));
 }
 
+/*
+ * Inline helper function to check a string ending for a match
+ */
+inline static bool endsWith(const std::string& str, const std::string& suffix) {
+  return (str.size() >= suffix.size()) &&
+          (0 == str.compare(str.size()-suffix.size(), suffix.size(), suffix));
+}
```
Any reason not to use StringEndsWith?:
https://github.com/OpenLightingProject/ola/blob/0.10/include/ola/StringUtils.h#L91-L97

kripton replied:
No reason, I was just not aware. Thanks for pointing it out to me, changed that :+1:

peternewman commented on common/file/UtilTest.cpp:
```diff
@@ -90,3 +100,54 @@ void UtilTest::testFilenameFromPath() {
+  bool rdm_test_server_found = 0;
+
+  for (const auto& value : files) {
+    if (endsWith(value, "rdm_model_collector.py.1"))
+      rdm_model_collector_found = 1;
+    if (endsWith(value, "rdm_responder_test.py.1"))
+      rdm_responder_test_found = 1;
+    if (endsWith(value, "rdm_test_server.py.1"))
+      rdm_test_server_found = 1;
```
Can we have { } brackets added please.

kripton replied:
sure, done

peternewman commented on common/file/UtilTest.cpp:
```diff
@@ -90,3 +100,54 @@ void UtilTest::testFilenameFromPath() {
+
+  for (const auto& value : files) {
+    if (endsWith(value, "rdm_model_collector.py.1"))
+      rdm_model_collector_found = 1;
```
Probably redundant, but is it worth adding some tests that `rdm_model_collector_found == 0` before setting it, to ensure we're not returning duplicates, and someone doesn't fix the test in some weird way in future, like by just bumping the expected count?

kripton replied:
Yes, I guess that would be redundant since it's a locally-scoped variable. But sure, it cannot hurt to assert that

peternewman replied:
I don't think it would be redundant if we returned ["rdm_model_collector.py.1", "rdm_model_collector.py.1",...] for example.

Thanks for fixing anyway.

peternewman commented:
Apologies for the flurry of commits @kripton but I've now got Codespell working again on Travis which would need to be sorted to merge this anyway, hope you don't mind.

kripton commented:
No need to excuse for the Travis-fixing commits.
I've taken care of your comments.
About your comment on:
> I'm unclear what a successful opendir call and an errorno value would mean, although it seems to say it shouldn't.

As far as I understand: If opendir fails, it'll return `NULL` and set `errno` so one can see why it failed. If it succeeds, it returns a value and probably leaves `errno` untouched. However, it could also set a value. But since there is no `errno`-value for "success", it probably doesn't. So if a value is returned, just ignore `errno`. I wouldn't know which value to ASSERT here. But before calling `readdir`, I set it to `0` to make sure it has a defined value.
I agree that a successful opendir WITH an value in errno would mean. But ASSERTING it to be 0 when the manuals says it could be changed seems a bit risky to me.

peternewman commented on common/file/UtilTest.cpp:
```diff
@@ -90,3 +94,56 @@ void UtilTest::testFilenameFromPath() {
+    if (ola::StringEndsWith(value, "rdm_model_collector.py.1")) {
+      OLA_ASSERT_FALSE(rdm_model_collector_found);  // make sure it's 0 before
+      rdm_model_collector_found = 1;
+    }
+    if (ola::StringEndsWith(value, "rdm_responder_test.py.1")) {
```
We could now make this an `else if` which would be slightly more performant.

peternewman commented on common/file/UtilTest.cpp:
```diff
@@ -90,3 +94,56 @@ void UtilTest::testFilenameFromPath() {
+    if (ola::StringEndsWith(value, "rdm_responder_test.py.1")) {
+      OLA_ASSERT_FALSE(rdm_responder_test_found);  // make sure it's 0 before
+      rdm_responder_test_found = 1;
+    }
+    if (ola::StringEndsWith(value, "rdm_test_server.py.1")) {
```
And here.

peternewman commented on common/file/Util.cpp:
```diff
@@ -151,8 +152,10 @@ bool FindMatchingFiles(const string &directory,
     }
-    if (readdir_r(dp, &dir_ent, &dir_ent_p)) {
-      OLA_WARN << "readdir_r(" << directory << "): " << strerror(errno);
+    errno = 0;
```
Can we just have a comment here then, explaining why we're setting it to zero (so no-one tidies it away in future).

kripton commented:
Hi @peternewman , I've added comments regarding `errno = 0`, changed the `if`s to `else if`s and modified `NEWS` and `AUTHORS`. Feel free to modify the two latter files if you prefer a different wording :)

peternewman commented on common/file/UtilTest.cpp:
```diff
@@ -90,3 +94,54 @@ void UtilTest::testFilenameFromPath() {
+      rdm_model_collector_found = 1;
+    } else if (ola::StringEndsWith(value, "rdm_responder_test.py.1")) {
+      OLA_ASSERT_FALSE(rdm_responder_test_found);  // make sure it's 0 before
+      rdm_responder_test_found = 1;
```
Here

peternewman commented on common/file/UtilTest.cpp:
```diff
@@ -90,3 +94,54 @@ void UtilTest::testFilenameFromPath() {
+      rdm_responder_test_found = 1;
+    } else if (ola::StringEndsWith(value, "rdm_test_server.py.1")) {
+      OLA_ASSERT_FALSE(rdm_test_server_found);  // make sure it's 0 before
+      rdm_test_server_found = 1;
```
Plus this one.

kripton commented on common/file/UtilTest.cpp:
```diff
@@ -90,3 +94,54 @@ void UtilTest::testFilenameFromPath() {
+void UtilTest::testFindMatchingFiles() {
+  bool okay = false;
+
+  std::vector<std::string> prefixes;
+  prefixes.push_back("rdm");
```
True. I first meant to test the function taking the vector but then realized that the string-based method just calls the vector-based version and as such I wrote the test only for the one calling the other. I'll remove that

peternewman replied:
It's probably overkill doing both as you say, especially looking at how it's implemented, there aren't corner cases for a second prefix, aside from a file being listed twice but that's likely user error effectively.

kripton commented on common/file/UtilTest.cpp:
```diff
@@ -90,3 +94,54 @@ void UtilTest::testFilenameFromPath() {
+  OLA_ASSERT_EQ_MSG(3, (int)files.size(),
+                    "Not exactly 3 files man/rdm_* returned");
+
+  bool rdm_model_collector_found = 0;
+  bool rdm_responder_test_found = 0;
+  bool rdm_test_server_found = 0;
```
You're right, `true` and `false` make more sense here. Changed :)

peternewman replied:
Erm thanks, but you haven't pushed...

kripton replied:
True, I haven't yet finished all commits I want to have in :) Working on it

kripton commented:
One thing I noticed: Travis build on macOS fails due to newer libmicrohttpd. I'm experiencing the same on my Gentoo Linux machine. v0.9.70 is fine, v0.9.71 fails.

peternewman commented:
> One thing I noticed: Travis build on macOS fails due to newer libmicrohttpd. I'm experiencing the same on my Gentoo Linux machine. v0.9.70 is fine, v0.9.71 fails.

Oops, that's embarrassing, I'd just assumed it was the tests that are always flaky on Mac and hadn't looked yet.

Looks like it's someone else who doesn't understand semver! See for example https://bugs.archlinux.org/task/67208 which links to https://git.gnunet.org/libmicrohttpd.git/commit/?id=6347f514aa2388e774d5bf356df8046864e5f73c .

Do you want to chuck all the necessary #ifdef cotton wool in, either in this PR or another one?

kripton commented:
Yep, that libmicrohttpd-changes are a pity. I'll create a separate PR for that, feel free to create an issue to track the build breakage.

kripton commented:
@peternewman I think I've done all changes + pushed now :) Let's see what Travis says, despite the expected-for-now failures on macOS

kripton commented:
#1650 created to track the libmicrohttpd build failures. However, now Travis fails on macOS due to some python file conflicts.....

peternewman commented:
> However, now Travis fails on macOS due to some python file conflicts.....

From what I can see, it's trying to upgrade the Homebrew packages, so postgis pulls in numpy via Homebrew but because numpy is already installed via pip it fails. Hopefully Travis will get their act together soon, or I can force merge it, or manually delete files/uninstall from pip or something...
